0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
+pushing this philosophy even further.
+
+The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
+pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
+a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
+sums the indices of all non-zero elements of the array `a`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Tracking the values enumerated so far and/or the number of iterations in a `foreach` loop
+is possible with manual helper functions as illustrated above, but only when the source collection is a sequence.
+It may be a good idea in the next iteration of the feature to have convenient mechanisms for this,
+if they are frequently requested. They may be best addressed when adding support for 
+user-defined collection types and enumerators instead. Another possibility is supporting
+parallel orderings as in Cryptol or Haskell, such that indexing an arbitrary collections could be expressed as
+`seq x <- s (some symbol) i <- (0 to *) :: (s, i)`, but I don't see an obvious choice of syntax.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Adding this building block to the Dafny language opens up a wide array of tempting new features!
+
+## Pattern matching on quantifier variables
+
+The LHS of `<-` declarations could be extended to allow case patterns, just as variable declaration statements do.
+This would allow destructuring datatypes and tuples, as in `(k, v) <- myMap.Items`.
+It could also support filtering, as in `Some(x) <- mySetOfOptionals`.
+This would be consistent with how case patterns are used in match statements; these two loops would be equivalent, for example:
+
+```dafny
+foreach Some(x) <- [None, Some(42), None] {
+  print x;
+}
+
+foreach o <- [None, Some(42), None] {
+  match o {
+    case Some(x) => print x;
+  }
+}
+```
+
+## User-defined collections
+
+As discussed in the earlier [`for` loops RFC](https://github.com/dafny-lang/rfcs/pull/4), the semantics of
+`for` or `foreach` loops in many programming languages are defined in terms of "iterator" or "enumerator" objects. 
+The features proposed here only support builtin collections, but could easily be extended to support
+arbitrary Dafny types that provide a way to obtain an enumerator for their contents.
+[This PR](https://github.com/dafny-lang/libraries/pull/37) should be a good start towards designing
+the necessary interface. This is also likely the best way to provide a better user experience
+for looping or quantifying over the values produced by an `iterator` type.
+
+## Range expressions
+
+This would interpret the existing `i..j` syntax, 
+currently only usable in sequence or array indexing expressions such as `s[i..j]`,
```

Comment:
Unless range expressions would be allowed everywhere and have their own type (which I resist at the moment), then ranges expressions would be allowed only to the right of `<-`. If so, I like the syntax @MikaelMayer suggested above, namely `x <- i to j`, which looks like in a `for` loop.